### PR TITLE
Add support for multiple listening ports

### PR DESCRIPTION
**Pull request by pschorf**
Related [Jet PR](https://github.com/twosigma/jet/pull/38)

## Changes proposed in this PR
- Add support for listening on multiple ports

## Why are we making these changes?
Allow waiter to listen on multiple ports.

### CODE REVIEW

File: waiter/src/waiter/settings.clj [hunk 1]
```diff
@@ -86,7 +86,8 @@
                                               s/Keyword schema/require-symbol-factory-fn}
                                              schema/contains-kind-sub-map?)
    ;; TODO port belongs in server-options?
-   (s/required-key :port) schema/positive-int
+   (s/optional-key :port) schema/positive-int
+   (s/required-key :ports) [schema/positive-int]
```
Comment (shamsimam) [inline waiter/src/waiter/settings.clj]: Should we move this to `server-options`?
Else, should we treat port as `hostname` where it can be a scalar or vector?

Comment (sradack) [reply waiter/src/waiter/settings.clj]: And we should do the same for Jet.  Dynamic typing FTW

Comment (pschorf) [reply waiter/src/waiter/settings.clj]: Keep `port` or rename to `ports`?

Comment (shamsimam) [reply waiter/src/waiter/settings.clj]: Keep `port` for backwards compatibility?

Comment (pschorf) [reply waiter/src/waiter/settings.clj]: Done

File: waiter/src/waiter/settings.clj [hunk 2]
```diff
@@ -86,7 +86,8 @@
                                               s/Keyword schema/require-symbol-factory-fn}
                                              schema/contains-kind-sub-map?)
    ;; TODO port belongs in server-options?
-   (s/required-key :port) schema/positive-int
+   (s/required-key :port)  (s/either schema/positive-int
```
Comment (shamsimam) [inline waiter/src/waiter/settings.clj]: Nitpick: Extra space between the closing and opening parenthesis.

File: waiter/src/waiter/main.clj [hunk 3]
```diff
@@ -77,7 +77,8 @@
    :state core/state
    :http-server (pc/fnk [[:curator kv-store]
                          [:routines generate-log-url-fn waiter-request?-fn websocket-request-acceptor]
```
Comment (shamsimam) [inline waiter/src/waiter/main.clj]: Nitpick: Unnecessary change for this PR.

**Comment by shamsimam** [comment]
Has the jet PR been updated?

File: waiter/src/waiter/settings.clj [hunk 4]
```diff
@@ -316,7 +317,7 @@
    :password-store-config {:kind :configured
                            :configured {:factory-fn 'waiter.password-store/configured-provider
                                         :passwords ["open-sesame"]}}
-   :port 9091
+   :ports []
```
[the remark this answers is not in the document]
Comment (pschorf) [reply waiter/src/waiter/settings.clj]: I think this was just out of date.

File: waiter/src/waiter/core.clj [hunk 5]
```diff
@@ -1266,9 +1273,9 @@
                                                                           instance-request-properties determine-priority-fn ws/process-response!
                                                                           ws/abort-request-callback-factory local-usage-agent request))]
                                      (->> process-request-fn
-                                       ws/wrap-ws-close-on-error
-                                       wrap-descriptor-fn
-                                       (ws/make-request-handler password))))
+                                          ws/wrap-ws-close-on-error
```
[the remark this answers is not in the document]
Comment (pschorf) [reply waiter/src/waiter/core.clj]: Hopefully fixed

File: waiter/src/waiter/core.clj [hunk 6]
```diff
@@ -1311,8 +1318,8 @@
                              (wrap-secure-request-fn
                                (fn ping-service-handler [request]
                                  (-> request
-                                   (update :headers assoc "x-waiter-fallback-period-secs" "0")
-                                   (handler))))))
+                                     (update :headers assoc "x-waiter-fallback-period-secs" "0")
```
[the remark this answers is not in the document]
Comment (pschorf) [reply waiter/src/waiter/core.clj]: Hopefully fixed

File: waiter/src/waiter/core.clj [hunk 7]
```diff
@@ -1335,15 +1342,15 @@
...
-                                     pr/wrap-too-many-requests
-                                     pr/wrap-suspended-service
-                                     pr/wrap-response-status-metrics
-                                     (interstitial/wrap-interstitial interstitial-state-atom)
-                                     wrap-descriptor-fn
-                                     wrap-secure-request-fn
-                                     wrap-auth-bypass-fn
-                                     wrap-https-redirect-fn
-                                     wrap-service-discovery-fn)))
+                                       pr/wrap-too-many-requests
```
[the remark this answers is not in the document]
Comment (pschorf) [reply waiter/src/waiter/core.clj]: IntelliJ 🖕

File: waiter/src/waiter/core.clj [hunk 8]
```diff
@@ -1687,9 +1694,9 @@
                                (fn wrap-secure-request-fn
                                  [handler]
                                  (let [handler (-> handler
-                                                 (cors/wrap-cors-request
-                                                   cors-validator waiter-request?-fn exposed-headers)
-                                                 authentication-method-wrapper-fn)]
+                                                   (cors/wrap-cors-request
```
[the remark this answers is not in the document]
Comment (pschorf) [reply waiter/src/waiter/core.clj]: Fixed?

File: waiter/src/waiter/main.clj (hunk 3 above)
Comment (pschorf) [reply waiter/src/waiter/main.clj]: 🤷‍♂ fixed

File: waiter/src/waiter/settings.clj (hunk 2 above)
Comment (pschorf) [reply waiter/src/waiter/settings.clj]: Done

**Comment by pschorf** [comment]
I removed the comment from the Jet PR

**Comment by pschorf** [comment]
@shamsimam build is green
